### PR TITLE
[Rebase & FF] 202405: DebugLib: add new assert behaviors 

### PR DESCRIPTION
**Pull request by VivianNK**
## Description

New AssertLib behavior:

- UEFI_810
- Add CpuBreakAssert

---

Changes from the following commits:

d3138534f6
93260d499b - split
411950837b

Removed definition of the following custom PI Status Codes from original commits:
- `EFI_SW_EC_MEMORY_TYPE_INFORMATION_CHANGE` 
  - PR to remove its usage here: https://github.com/microsoft/mu_plus/pull/524
- `EFI_SW_EC_RELEASE_ASSERT` 
  - PR to remove its usage here: https://github.com/microsoft/mu_plus/pull/530

---
- [ ] Impacts functionality?
- [ ] Impacts security?
- [x] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested

Changes from release/202311

## Integration Instructions
Upgrading from edk2  -> 202405:
No change for platform

Upgrading from 202311 -> 202405:
Change any reference to the PI Status Code macros as follows:

- Replace `EFI_SW_EC_MEMORY_TYPE_INFORMATION_CHANGE` Status Code with `EFI_SW_EC_INCONSISTENT_MEMORY_MAP`
- Replace `EFI_SW_EC_RELEASE_ASSERT` Status Code with `EFI_SW_EC_ILLEGAL_SOFTWARE_STATE`

### CODE REVIEW

File: MdePkg/Include/Pi/PiStatusCode.h [hunk 1]
```diff
@@ -966,27 +966,29 @@ typedef struct {
...
+#define EFI_SW_EC_CFG_CLR_REQUEST                 0x0000000C
+#define EFI_SW_EC_CFG_DEFAULT                     0x0000000D
+#define EFI_SW_EC_PWD_INVALID                     0x0000000E
+#define EFI_SW_EC_PWD_CLR_REQUEST                 0x0000000F
+#define EFI_SW_EC_PWD_CLEARED                     0x00000010
+#define EFI_SW_EC_EVENT_LOG_FULL                  0x00000011
+#define EFI_SW_EC_WRITE_PROTECTED                 0x00000012
+#define EFI_SW_EC_FV_CORRUPTED                    0x00000013
+#define EFI_SW_EC_INCONSISTENT_MEMORY_MAP         0x00000014
+#define EFI_SW_EC_MEMORY_TYPE_INFORMATION_CHANGE  0x00000015 // MU_CHANGE
```
Comment (kuqin12) [inline MdePkg/Include/Pi/PiStatusCode.h]: I thought this is the same as `EFI_SW_EC_INCONSISTENT_MEMORY_MAP`? Do we need this 0x15 definition for anything?

File: MdePkg/Include/Library/DebugLib.h [hunk 2]
```diff
@@ -19,12 +19,14 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
...
-#define DEBUG_PROPERTY_CLEAR_MEMORY_ENABLED       0x08
-#define DEBUG_PROPERTY_ASSERT_BREAKPOINT_ENABLED  0x10
-#define DEBUG_PROPERTY_ASSERT_DEADLOOP_ENABLED    0x20
+#define DEBUG_PROPERTY_DEBUG_ASSERT_ENABLED        0x01
+#define DEBUG_PROPERTY_DEBUG_PRINT_ENABLED         0x02
+#define DEBUG_PROPERTY_DEBUG_CODE_ENABLED          0x04
+#define DEBUG_PROPERTY_CLEAR_MEMORY_ENABLED        0x08
+#define DEBUG_PROPERTY_ASSERT_BREAKPOINT_ENABLED   0x10
+#define DEBUG_PROPERTY_ASSERT_DEADLOOP_ENABLED     0x20
+#define DEBUG_PROPERTY_ASSERT_BREAKASSERT_ENABLED  0x40 // MU_CHANGE BREAKASSERT
```
Comment (kuqin12) [inline MdePkg/Include/Library/DebugLib.h:28]: I think the baselib and debuglib changes should be broken up into 2 commits to keep their own sanity?

**Review by kuqin12 — DISMISSED** [review]

File: MdePkg/Include/Pi/PiStatusCode.h (hunk 1 above)
Comment (VivianNK) [reply MdePkg/Include/Pi/PiStatusCode.h]: Looks like `EFI_SW_EC_MEMORY_TYPE_INFORMATION_CHANGE` is used here for MemoryTypeInformationChangeLib: https://github.com/microsoft/mu_plus/blob/005cfdd5b9942e98ff6218a7b06360f6b7606396/MsCorePkg/Library/MemoryTypeInformationChangeLib/MemoryTypeInformationChangeLib.c#L38

Which is not what EFI_SW_EC_INCONSISTENT_MEMORY_MAP  is defined for.

From the PI Spec:

> EFI_SW_EC_INCONSISTENT_MEMORY_MAP 
> System will reboot due to inconsistent memory
maps

> 0x0015-0x00FF
> Reserved for future use by this specification for Host Software class error codes.

>0x8000-0xFFFF 
>Reserved for OEM use.

So ideally 0x15 wouldn't be defined here but it is already being consumed for telemetry. https://github.com/microsoft/mu_basecore/blob/7cf3382ebaf5b5b5a2487d7d13e910c1c7dac233/MdeModulePkg/Library/UefiBootManagerLib/BmMisc.c#L329

I can break it out into its own commit since it seems unrelated to the addition of
> EFI_SW_EC_RELEASE_ASSERT

What do you think?

File: MdePkg/Include/Library/DebugLib.h (hunk 2 above)
Comment (VivianNK) [reply MdePkg/Include/Library/DebugLib.h:28]: Split into 3 commits, an additional one for the new Pi Status codes. though re: your other comment, one of the new Pi codes does not seem related.

File: MdePkg/Include/Pi/PiStatusCode.h (hunk 1 above)
Comment (VivianNK) [reply MdePkg/Include/Pi/PiStatusCode.h]: No where in edk2 or in MU repos is EFI_SW_EC_INCONSISTENT_MEMORY_MAP consumed, interestingly.

Comment (kuqin12) [reply MdePkg/Include/Pi/PiStatusCode.h]: Yeah, the MU_CHANGE definition `EFI_SW_EC_MEMORY_TYPE_INFORMATION_CHANGE` should be replaced with `EFI_SW_EC_INCONSISTENT_MEMORY_MAP`. This macro was renamed during the process of updating the PI spec...

Comment (VivianNK) [reply MdePkg/Include/Pi/PiStatusCode.h]: Talked offline, `EFI_SW_EC_MEMORY_TYPE_INFORMATION_CHANGE` is defunct and can be removed. Will need to remove reference from MU_PLUS

File: MdePkg/Library/BaseLib/AArch64/CpuBreakAssert.S [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# @file CpuBreakAssert.S
+# 
+# CpuBreakAssert function for AArch64.
+# 
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (os-d) [inline MdePkg/Library/BaseLib/AArch64/CpuBreakAssert.S]: Copyright (c) Microsoft Corporation.

For all files.

File: MdePkg/Library/BaseLib/BaseLib.inf [hunk 4]
```diff
@@ -226,6 +226,7 @@
   X64/Lfence.nasm
 
   X64/CpuBreakpoint.c | MSFT
+  X64/CpuBreakAssert.c | MSFT ## MS_CHANGE
```
Comment (os-d) [inline MdePkg/Library/BaseLib/BaseLib.inf]: Change to MU_CHANGE, for all files, please.

File: MdePkg/Include/Library/DebugLib.h [hunk 5]
```diff
@@ -19,12 +19,14 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
...
+#define DEBUG_PROPERTY_ASSERT_TELEMETRY_ENABLED    0x80 // MU_CHANGE LOGTELEMETRY
```
Comment (os-d) [inline MdePkg/Include/Library/DebugLib.h:29]: This this be coming in a different commit?

**Review by os-d — APPROVED** [review]

File: MdePkg/Include/Library/DebugLib.h (hunk 5 above)
Comment (VivianNK) [reply MdePkg/Include/Library/DebugLib.h:29]: Are you saying this (`DEBUG_PROPERTY_ASSERT_TELEMETRY_ENABLED`) should be in a separate commit/split it in two?

Comment (os-d) [reply MdePkg/Include/Library/DebugLib.h:29]: @kuqin12 , LOGTELEMETRY doesn't get used anywhere, is that because the expectation is for a platform to implement it?

Comment (VivianNK) [reply MdePkg/Include/Library/DebugLib.h:29]: @os-d This macro is used in AdvLoggerPkg: https://github.com/microsoft/mu_plus/blob/90f8ca98e2a7c7bc22a60b6e9fe18c69ce57d92f/AdvLoggerPkg/Library/AssertTelemetryLib/AssertLib.c#L95

**Review by apop5 — APPROVED** [review]

**Comment by VivianNK** [comment]
Merge after https://github.com/microsoft/mu_plus/pull/524

File: MdePkg/Library/UefiDebugLibDebugPortProtocol/DebugLib.c [hunk 6]
```diff
@@ -280,8 +280,15 @@ DebugAssert (
     UefiDebugLibDebugPortProtocolWrite (Buffer, AsciiStrLen (Buffer));
 
     //
-    // Generate a Breakpoint, DeadLoop, or NOP based on PCD settings
+    // Generate an Assertion Break, Breakpoint, DeadLoop, or NOP based on PCD settings
     //
+    // MU_CHANGE BEGIN BREAKASSERT
```
Comment (makubacki) [inline MdePkg/Library/UefiDebugLibDebugPortProtocol/DebugLib.c]: Technically, the MU_CHANGE should include the modification to the comment but not a big deal.

File: MdePkg/Include/Pi/PiStatusCode.h
Comment (makubacki) [inline MdePkg/Include/Pi/PiStatusCode.h:1]: I think this commit message in particular could use an explanation of the use case for the bit to reduce ambiguity and potential misuse.

**Review by makubacki — APPROVED** [review]

File: MdePkg/Include/Pi/PiStatusCode.h
Comment (VivianNK) [reply MdePkg/Include/Pi/PiStatusCode.h:1]: It's used in AdvLogger's AssertLib https://github.com/microsoft/mu_plus/blob/cdb7c26f948748fd5932bf8ddf98ffa2ff9d1cd7/AdvLoggerPkg/Library/AssertTelemetryLib/AssertLib.c#L113

Do you have any more details on the extent of the intended use case? Is it exclusively for asserts raised by DebugLib?

File: MdePkg/Library/UefiDebugLibDebugPortProtocol/DebugLib.c (hunk 6 above)
Comment (VivianNK) [reply MdePkg/Library/UefiDebugLibDebugPortProtocol/DebugLib.c]: updated

File: MdePkg/Include/Pi/PiStatusCode.h
Comment (kuqin12) [reply MdePkg/Include/Pi/PiStatusCode.h:1]: Per offline discussion, this should not be brought in anymore but instead replaced by `EFI_SW_EC_ILLEGAL_SOFTWARE_STATE`, right?

**Review by kuqin12 — APPROVED** [review]

File: MdePkg/Include/Pi/PiStatusCode.h
Comment (VivianNK) [reply MdePkg/Include/Pi/PiStatusCode.h:1]: @kuqin12 I thought that was only in regards to `EFI_SW_EC_INCONSISTENT_MEMORY_MAP`. Is it also true that `EFI_SW_EC_RELEASE_ASSERT` should be replaced with `EFI_SW_EC_ILLEGAL_SOFTWARE_STATE` in AssertLib?

Comment (kuqin12) [reply MdePkg/Include/Pi/PiStatusCode.h:1]: EFI_SW_EC_INCONSISTENT_MEMORY_MAP is to replace the mu definition 0x15 :) 0x16 is to be replaced by illegal software state :)

Comment (VivianNK) [reply MdePkg/Include/Pi/PiStatusCode.h:1]: Removed `EFI_SW_EC_RELEASE_ASSERT `
